pmlc/conversion/gpu/gpu_to_vulkan.cc

```diff
@@ -0,0 +1,466 @@
+    // Create LLVM constant for the descriptor binding index.
+    Value descriptorBinding = builder.create<LLVM::ConstantOp>(
+        loc, getLLVMInt32Type(), builder.getI32IntegerAttr(bindIndex));
+    if (auto memRefType = buffer.getType().dyn_cast_or_null<MemRefType>()) {
+      auto rank = memRefType.getRank();
+      auto elementType = memRefType.getElementType();
+      auto dynamicBuffer = builder.create<mlir::MemRefCastOp>(
+          loc, buffer, getDynamicMemRefType(rank, elementType));
+      Value unrankedBuffer = builder.create<mlir::MemRefCastOp>(
+          loc, dynamicBuffer, getUnrankedMemRefType(elementType));
+      builder.create<CallOp>(
+          loc, ArrayRef<Type>{},
+          builder.getSymbolRefAttr(getBufferBindingFunc(elementType)),
+          ArrayRef<Value>{vulkanRuntime, descriptorSet, descriptorBinding,
+                          unrankedBuffer});
+    } else {
+      return failure();
+    }
+  }
+  return success();
+}
+
+LogicalResult ConvertGpuLaunchFuncToVulkanCalls::transferBuffers(
+    Location loc, OpBuilder &builder, gpu::LaunchFuncOp launchOp) {
+  auto buffers = launchOp.operands();
+  for (size_t i = 0; i < buffers.size(); i++) {
+    for (auto pair : bufferMap) {
+      if (pair.first == buffers[i]) {
+        Value dst_index = builder.create<LLVM::ConstantOp>(
+            loc, getLLVMInt64Type(), builder.getI64IntegerAttr(lauchFuncIndex));
+        Value dst_binding = builder.create<LLVM::ConstantOp>(
+            loc, getLLVMInt64Type(), builder.getI64IntegerAttr(i));
+        Value src_index = builder.create<LLVM::ConstantOp>(
+            loc, getLLVMInt64Type(), builder.getI64IntegerAttr(pair.second[0]));
+        Value src_binding = builder.create<LLVM::ConstantOp>(
+            loc, getLLVMInt64Type(), builder.getI64IntegerAttr(pair.second[1]));
+
+        builder.create<LLVM::CallOp>(
+            loc, ArrayRef<Type>{},
+            builder.getSymbolRefAttr(kCreateVulkanMemoryTransferAction),
+            ArrayRef<Value>{vulkanRuntime, src_index, src_binding, dst_index,
+                            dst_binding});
+      }
+    }
+    llvm::SmallVector<uint64_t, 2> second;
+    second.append({lauchFuncIndex, i});
+    bufferMap[buffers[i]] = second;
+  }
+  return success();
+}
+
+LogicalResult ConvertGpuLaunchFuncToVulkanCalls::printBuffer(Location loc,
+                                                             OpBuilder &builder,
+                                                             Value &buffer) {
+  auto type = buffer.getType();
+  if (auto memRefType = type.dyn_cast_or_null<MemRefType>()) {
+    auto rank = memRefType.getRank();
+    auto elementType = memRefType.getElementType();
+    if (elementType.isF32()) {
+      auto dynamicBuffer = builder.create<mlir::MemRefCastOp>(
+          loc, buffer, getDynamicMemRefType(rank, elementType));
+      auto unrankedBuffer = builder.create<mlir::MemRefCastOp>(
+          loc, dynamicBuffer, getUnrankedMemRefType(elementType));
```

Comment:
Can we cast directly to an unranked type?

Reply:
Yes. But one thing is that the mlir llvm lowing pass does not implement unranked memref types for function signatures, so I added that in pmlc/conversion/gpu/llvm_lowering.cc, converting unranked memref to a rank variable and a pointer to descriptor and then it works.

Reply:
I'm not sure why you think that, because we use unranked memrefs in other places at the LLVM level.

Reply:
I think we can go from a static ranked memref to an unranked memref in a single cast.